### PR TITLE
Deprecate zonejs dependency

### PR DESCRIPTION
**Pull request by markwolff**
- Deprecate zonejs and switch to [cls-hooked](https://github.com/Jeff-Lewis/cls-hooked)
- add [`startOperation`](https://github.com/Microsoft/ApplicationInsights-node.js/issues/452#issuecomment-438383610)

### CODE REVIEW

File: AutoCollection/CorrelationContextManager.ts [hunk 1]
```diff
@@ -39,6 +41,8 @@ export interface CorrelationContext {
 export class CorrelationContextManager {
     private static enabled: boolean = false;
     private static hasEverEnabled: boolean = false;
+    private static session: any;
+    private static cls: any;
```
Comment (OsvaldoRosado) [inline AutoCollection/CorrelationContextManager.ts]: You can use the import style seen in the top of `https://github.com/Microsoft/ApplicationInsights-node.js/blob/develop/AutoCollection/diagnostic-channel/initialization.ts` to get the proper types for this without requiring the cls-hooked package in the resulting JS.

File: AutoCollection/CorrelationContextManager.ts [hunk 2]
```diff
@@ -49,7 +53,12 @@ export class CorrelationContextManager {
         if (!CorrelationContextManager.enabled) {
             return null;
         }
-        return Zone.current.get("context");
+        const context = CorrelationContextManager.session.get('context');
```
Comment (OsvaldoRosado) [inline AutoCollection/CorrelationContextManager.ts]: I wonder if we should choose a more AI-specific name than just `context` for this. I know it's also what we used in Zone, and it didn't seem to cause trouble there, but I'm thinking its possible to have conflicts if we're instrumenting an app that already uses cls.

File: AutoCollection/CorrelationContextManager.ts [hunk 3]
```diff
@@ -116,30 +124,23 @@ export class CorrelationContextManager {
...
-                // averting require's cache somehow.
-                if (typeof Zone === "undefined") {
-                    require("zone.js");
+                if (typeof this.cls === "undefined") {
+                    this.cls = require('cls-hooked');
                 }
             } catch (e) {
-                // Zone was already loaded even though we couldn't find its global variable
-                Logging.warn("Failed to require zone.js");
+                // cls-hooked was already loaded even though we couldn't find its global variable
```
Comment (OsvaldoRosado) [inline AutoCollection/CorrelationContextManager.ts]: It looks like cls-hooked doesn't use a global ambient variable like Zone did. This try/catch might not be necessary. Even if we keep it, the comment is now slightly wrong :smile:

File: AutoCollection/CorrelationContextManager.ts [hunk 4]
```diff
@@ -116,30 +124,23 @@ export class CorrelationContextManager {
...
+                Logging.warn("Failed to require cls-hooked");
             }
 
+            CorrelationContextManager.session = this.cls.createNamespace("AI-CLS-Session");
```
Comment (OsvaldoRosado) [inline AutoCollection/CorrelationContextManager.ts]: Out of curiosity, what does this do?

Comment (OsvaldoRosado) [reply AutoCollection/CorrelationContextManager.ts]: Also note, you're outside of the try/catch here. this.cls might be null.

**Comment by OsvaldoRosado** [comment]
Should our HTTP patching be updated as well to call `bindEmitter`?

File: AutoCollection/CorrelationContextManager.ts [hunk 5]
```diff
@@ -161,137 +161,6 @@ export class CorrelationContextManager {
         var nodeVer = process.versions.node.split(".");
         return parseInt(nodeVer[0]) > 3 || (parseInt(nodeVer[0]) > 2 && parseInt(nodeVer[1]) > 2);
```
Comment (OsvaldoRosado) [inline AutoCollection/CorrelationContextManager.ts]: This version check needs to be bumped higher. [package.json of `cls-hooked`](https://github.com/Jeff-Lewis/cls-hooked/blob/master/package.json) says >= 4.7.0

File: AutoCollection/CorrelationContextManager.ts [hunk 6]
```diff
@@ -97,6 +97,13 @@ export class CorrelationContextManager {
         }
     }
 
+    /**
+     * Wrapper for cls-hooked bindEmitter method
+     */
+    public static bindEmitter(reqres: any): void {
```
Comment (OsvaldoRosado) [inline AutoCollection/CorrelationContextManager.ts]: Should we name this `wrapEmitter` to be consistent with `wrapCallback`? Also, shouldn't the type here be `EventEmitter`?

Comment (OsvaldoRosado) [reply AutoCollection/CorrelationContextManager.ts]: Minor nit: Would prefer a generic argument name like `emitter` to `reqres`

Comment (OsvaldoRosado) [reply AutoCollection/CorrelationContextManager.ts]: This also needs if guards to only act when this is enabled

File: AutoCollection/HttpRequests.ts [hunk 7]
```diff
@@ -90,6 +90,8 @@ class AutoCollectHttpRequests {
                     // Set up correlation context
                     const requestParser = new HttpRequestParser(request);
                     const correlationContext = this._generateCorrelationContext(requestParser);
+                    CorrelationContextManager.bindEmitter(request);
```
Comment (OsvaldoRosado) [inline AutoCollection/HttpRequests.ts]: I wonder if we should do this in the auto collector for HTTP dependencies as well.

File: tsconfig.json [hunk 8]
```diff
@@ -17,7 +17,7 @@
         "Tests/**/*.ts"
     ],
     "files": [
-        "./node_modules/zone.js/dist/zone.js.d.ts"
+        "./node_modules/@types/cls-hooked/index.d.ts"
```
Comment (OsvaldoRosado) [inline tsconfig.json]: Since cls' types don't appear to be ambient, you might not need this at all

File: .travis.yml [hunk 9]
```diff
@@ -21,6 +24,9 @@ before_script:
 script:
   - npm test
   # Only run funcs for LTS releases
+  - if [ "$TRAVIS_NODE_VERSION" = "11" ]; then npm run functionaltest; fi
```
Comment (OsvaldoRosado) [inline .travis.yml]: Nit: 11 isn't an LTS release. It's okay if we want to establish a new criteria for what belongs here, but the comment should be updated

File: .travis.yml [hunk 10]
```diff
@@ -21,6 +24,9 @@ before_script:
...
+  - if [ "$TRAVIS_NODE_VERSION" = "10" ]; then npm run functionaltest; fi
+  - if [ "$TRAVIS_NODE_VERSION" = "8.0.0" ]; then npm run functionaltest; fi
```
Comment (OsvaldoRosado) [inline .travis.yml]: I wonder if we should have a comment explaining why we add 8.0.0 in particular in addition to 8

File: .travis.yml (hunk 9 above)
Comment (markwolff) [reply .travis.yml]: updated to run func tests on current release as well

File: .travis.yml (hunk 10 above)
Comment (markwolff) [reply .travis.yml]: Added comment

File: Tests/FunctionalTests/Runner/TestValidation.js [hunk 11]
```diff
@@ -83,7 +83,7 @@ module.exports.TestValidation = class TestValidation {
                 // Did we find all of the items in the data set?
                 if (dataSet.length > 1){
                     Utils.Logging.error("FAILED EXPECTATION - Unexpected child telemetry item(s)!");
-                    Utils.Logging.error(JSON.stringify(dataSet));
+                    Utils.Logging.error(JSON.stringify(dataSet, null, 2));
```
Comment (OsvaldoRosado) [inline Tests/FunctionalTests/Runner/TestValidation.js]: What does this change do?

**Comment by OsvaldoRosado** [comment]
Not a blocker for this PR. But before a stable release we should expose a switch on `setAutoDependencyCorrelation` to force `continuation-local-storage` vs `cls-hooked`.

File: Tests/FunctionalTests/Runner/TestValidation.js (hunk 11 above)
Comment (markwolff) [reply Tests/FunctionalTests/Runner/TestValidation.js]: pretty prints the errors for my sake

**Pull request merged**